clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java

```diff
@@ -232,6 +232,15 @@ public RuntimeException lastError() {
         return lastError;
     }
 
+    public synchronized boolean ensurePartitionAdded(TopicPartition tp) {
+        if (isInTransaction() && !partitionsInTransaction.contains(tp)) {
+            transitionToFatalError(new IllegalStateException("Attempted to dequeue a record batch to send " +
```

Comment:
Would this ever happen assuming we do not have bugs? If yes then we should probably throw IllegalStateException directly to indicate a bug?